File: post-processing/plots.m
%% PLOTS
%
%   Reads the .h5 data file and plots the results
%
% (c) Matteo M. 2022

clear;


%% Settings
plot_profiles = 1;      % Plot profiles along the column
plot_bm_profiles = 1;   % Plot profiles along the biofilm
units = 2;              % 1: ppb, 2: ug/m3
ds = 10;                % Plot every ds samples

%% Open GUI
[FileName,PathName] = uigetfile('../results/*.h5','Select the .h5 data file');
filepath = [PathName FileName];

%% Read data
fid = H5F.open(filepath);
dset_id = H5D.open(fid,'/parameters/use_ts');
use_ts = boolean(H5D.read(dset_id));

dset_id = H5D.open(fid,'/parameters/mixing_model');
mixing_model = boolean(H5D.read(dset_id));

dset_id = H5D.open(fid,'/parameters/mw');
mw = double(H5D.read(dset_id));

dset_id = H5D.open(fid,'/parameters/filename');
yaml = H5D.read(dset_id);

dset_id = H5D.open(fid,'/parameters/export_profiles');
export_profiles = boolean(H5D.read(dset_id));

dset_id = H5D.open(fid,'/parameters/export_bm_profiles');
export_bm_profiles = boolean(H5D.read(dset_id));

% Conversion factor ppb -> ug/m3
if units == 2
    cf = mw/24.45;
else
    cf = 1;
end

if use_ts
    % Inlet
    inlet.data = h5read(filepath,'/data/inlet/c');
    inlet.t = h5read(filepath,'/data/inlet/t');

    % Outlet
    outlet.data = h5read(filepath,'/data/outlet/c');
    outlet.t = h5read(filepath,'/data/outlet/t');
end

% Model
model.c = h5read(filepath,'/results/output/c');
model.t = h5read(filepath,'/results/output/t');

if mixing_model
    model.mix = h5read(filepath,'/results/output/c_mix');
end

%% PLOTS
if use_ts
    % Output concentration
    figure;
    plot(double(inlet.t)/3600,inlet.data*cf,'-r')
    hold all
    plot(double(outlet.t)/3600,outlet.data*cf,'-b')
    plot(double(model.t(1:ds:end))/3600,model.data(1:ds:end)*cf, 'ok')
    xlabel('Time (h)')
    if units == 2
        ylabel('Concentration (\mug/m^3)')
    else
        ylabel('Concentration (ppb)')
    end
    legend('Inlet','Outlet','Model')
end

% Mixing model
if mixing_model
    figure;
    plot(double(model.t)/3600,model.c*cf, 'k')
    hold all
    plot(double(model.t)/3600,model.mix*cf)
    if units == 2
        ylabel('Concentration (\mug/m^3)')
    else
        ylabel('Concentration (ppb)')
    end
    legend('Column outlet','Inlet air - mixed')
end


% Column profiles
if plot_profiles && export_profiles
    profiles.t = h5read(filepath,'/results/profiles/t');
    profiles.z = h5read(filepath,'/results/profiles/z');
    profiles.o = h5read(filepath,'/results/profiles/o');
    profiles.c = h5read(filepath,'/results/profiles/c');
    
    % Oxygen
    figure;
    l = 1;
    for i = 1:length(profiles.t)
        plot(profiles.z,profiles.o(:,i))
        if i == 1
            hold all
        end
        lgnd_o{l} = ['t = ' num2str(profiles.t(i)/3600,'%.2f') ' h'];
        l = l+1;
    end
    xlabel('Distance from inlet (m)')
    ylabel('Oxygen concentration (ppm)')
    legend(lgnd_o)
    
    % VOC
    figure;
    l = 1;
    for i = 1:length(profiles.t)
        plot(profiles.z,profiles.c(:,i)*cf)
        if i == 1
            hold all
        end
        lgnd_i{l} = ['t = ' num2str(profiles.t(i)/3600,'%.2f') ' h'];
        l = l+1;
    end
    xlabel('Distance from inlet (m)')
    if units == 2
        ylabel('VOC concentration (\mug/m^3)')
    else
        ylabel('VOC concentration (ppb)')
    end
    legend(lgnd_i)
end

% Biofilm profiles
if plot_bm_profiles && export_bm_profiles
    profiles.t = h5read(filepath,'/results/profiles/t');
    profiles.x = h5read(filepath,'/results/profiles/x');
    profiles.uo = h5read(filepath,'/results/profiles/uo');
    profiles.ui = h5read(filepath,'/results/profiles/ui');

    % Oxygen
    figure;
    l = 1;
    for i = 1:length(profiles.t)
        plot(profiles.x,profiles.uo(:,i))
        if i == 1
            hold all
        end
        lgnd_o{l} = ['t = ' num2str(profiles.t(i)/3600,'%.2f') ' h'];
        l = l+1;
    end
    xlabel('Distance biofilm surface (\mum)')
    ylabel('Oxygen concentration in the biofilm (ppm)')
    legend(lgnd_o)
    
    % VOC
    figure;
    l = 1;
    for i = 1:length(profiles.t)
        plot(profiles.x,profiles.ui(:,i)*cf)
        if i == 1
            hold all
        end
        lgnd_i{l} = ['t = ' num2str(profiles.t(i)/3600,'%.2f') ' h'];
        l = l+1;
    end
    xlabel('Distance from biofilm surface (\mum)')
    if units == 2
        ylabel('VOC concentration in the biofilm (\mug/m^3)')
    else
        ylabel('VOC concentration in the biofilm (ppb)')
    end
    legend(lgnd_i)
end

File: post-processing/plots.py
""" 
PLOTS
======================

Reads .h5 files and plots the results 
    
(c) Matteo M. 2022

"""

import os, fnmatch
import numpy as np
import matplotlib.pyplot as plt
import h5py


##########   SETTINGS  #############
results_dir = '../results/'    # Results files location
plot_profiles = True           # Plot profiles along the column
skip_profiles = 0              # Skip n profiles
ds = 1                         # Plot every ds samples
units = 2                      # 1: ppb, 2:ug/m3
####################################


# List of files
def listfiles(directory):
    listOfFiles = os.listdir(directory)
    pattern = "*.h5"
    l = []
    for entry in listOfFiles:
        if fnmatch.fnmatch(entry, pattern):
            l.append(entry)
    if len(l) < 1:
        print('Error. No .h5 file found.')
    return l



# Print the list of files and select one from the list
l_files = listfiles(results_dir) 
num_files = len(l_files)
for i in range(num_files):
    print(f'[{i}] {l_files[i]}')
print(f'Select a file (default: last file) [{num_files-1}]:', end = '')
num = input()
if len(num) > 0:
    sel = int(num)
else:
    sel = num_files-1

filename = results_dir + l_files[sel]



########### READ .H5 FILES ############
hf = h5py.File(filename, 'r')

model = {}
model['t'] = hf['/results/output/t'][:]
model['c'] = hf['/results/output/c'][:]


use_ts = hf['parameters/use_ts']
if use_ts[()]:
    inlet = {}
    inlet['t'] = hf['data/inlet/t'][:]
    inlet['c'] = hf['data/inlet/c'][:]

    outlet = {}
    outlet['t'] = hf['data/outlet/t'][:]
    outlet['c'] = hf['data/outlet/c'][:]

export_profiles = hf['parameters/export_profiles']
if export_profiles[()]:
    profiles = {}
    profiles['t'] = hf['/results/profiles/t'][:]
    profiles['z'] = hf['/results/profiles/z'][:]
    profiles['o'] = hf['/results/profiles/o']
    profiles['c'] = hf['/results/profiles/c']

mixing_model = hf['parameters/mixing_model']
if mixing_model[()]:
    model['c_mix'] = hf['/results/output/c_mix'][:]
	
# Print simulation filename (.yml)
print(hf['parameters/filename'].asstr()[()])

	
################ PLOTS ################

# Read molar weight
mw = hf['parameters/mw'][()]

# Conversion factor
if units == 2:
    cf = mw/24.45
else:
    cf = 1


# Output concentrations
plt.figure()

if use_ts[()]:
    plt.plot(inlet['t']/3600,inlet['c']*cf,linestyle='-',marker='',fillstyle='none', label='Inlet')
    plt.plot(outlet['t']/3600,outlet['c'],linestyle='-',marker='',fillstyle='none', label='Outlet')

plt.plot(model['t'][::ds]/3600,model['c'][::ds]*cf,linestyle='none',marker='o',fillstyle='none', label='Column outlet')

if mixing_model[()]:
    plt.plot(model['t'][::ds]/3600,model['c_mix'][::ds]*cf,linestyle='none',marker='o',fillstyle='none', label='Inlet air - mixed')

plt.legend()
plt.xlabel('Time (h)')
if units == 2:
    plt.ylabel('Concentration ($\mu$g/m$^3$)')
else:
    plt.ylabel('Concentration (ppb)')


if (export_profiles[()] == 1) and plot_profiles:
    # Profiles
    # Oxygen
    plt.figure()

    for i in range(len(profiles['t'])):
        if i > skip_profiles:
            plt.plot(profiles['z'],profiles['o'][i][:], linestyle='-',marker='',fillstyle='none', label=f"t = {profiles['t'][i]/3600:.1f} h")
    plt.legend()
    plt.xlabel('Distance from inlet (m)')
    plt.ylabel('Oxygen concentration (ppm)')

    # VOC
    plt.figure()
    for i in range(len(profiles['t'])):
        if i > skip_profiles:
            plt.plot(profiles['z'],profiles['c'][i][:]*cf, linestyle='-',marker='',fillstyle='none', label=f"t = {profiles['t'][i]/3600:.1f} h")
    plt.legend()
    plt.xlabel('Distance from inlet (m)')
    if units == 2:
        plt.ylabel('Concentration ($\mu$g/m$^3$)')
    else:
        plt.ylabel('Concentration (ppb)')

plt.show()




